fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/PlanOrderingKey.java

```diff
@@ -65,13 +68,25 @@ public class PlanOrderingKey {
     private final int prefixSize;
     private final int primaryKeyStart;
     private final int primaryKeyTail;
+    // Keep a set of positions that are equality bound but are not in the prefix. (All components in the prefix are equality bound.)
+    @Nonnull
+    private final Set<Integer> equalityBoundInTail;
 
     public PlanOrderingKey(@Nonnull List<KeyExpression> keys, int prefixSize,
                            int primaryKeyStart, int primaryKeyTail) {
         this.keys = keys;
         this.prefixSize = prefixSize;
         this.primaryKeyStart = primaryKeyStart;
         this.primaryKeyTail = primaryKeyTail;
+
+        ImmutableSet.Builder<Integer> equalityBoundBuilder = ImmutableSet.builder();
+        for (int pos = prefixSize; pos < keys.size(); pos++) {
+            final KeyExpression component = keys.get(pos);
+            if (keys.indexOf(component) < prefixSize) {
+                equalityBoundBuilder.add(pos);
+            }
+        }
+        this.equalityBoundInTail = equalityBoundBuilder.build();
```

Comment:
Could you add some comments here? Also, you mentioned that `though the PR as written does not handle if the same field appears twice in the non-equality bound section, so I'm not sure we can close that issue out (yet)` in the PR summary part. My understanding is that if in an unnested query, if we `order by columnA, columnB`, we could have the same prefix field appears twice in the non-equality bound section? why doesn't the current modification solve this case?

Reply:
> My understanding is that if in an unnested query, if we order by columnA, columnB, we could have the same prefix field appears twice in the non-equality bound section?

In theory, yes, this could happen. It would require an index which contained the same component multiple times. I was thinking that this could happen if one of the primary key components was also in the index. (That's generally normalized out, though I thought the `ListKeyExpression` used by synthetic types was complicating that.)

> why doesn't the current modification solve this case?

I'm not sure if your question is "what about this code makes it so that this case isn't handled?" or "why did you decide to write code that didn't handle this case?".

If the former, the problem is that if the _first_ occurrence of a key expression occurs in the non-equality bound tail, then when we go to match it because we'll calculate the index of the position based on its _first_ occurrence, which won't match the expected `nextNonPrefix` value. That's not a problem for equality bound keys, because we check if the key is equality bound first.

If the latter, the reason I didn't handle it was because I was working off of an example where (I thought) a component was equality bound, and so I wanted to make sure that that case works. (As it happens, I'm actually less sure that that was necessary, because the real problem was that the equality-bound component was being normalized in a bit of weird way.) Keeping track of the way the tail works is a bit tricky, and so I didn't want to do that if I didn't develop tests for it. I started going about fixing it, and I think it might actually not be too bad (because, I think, it may be sufficient to check only that the component's position is `< nextNonPrefix`), but I would prefer to get this in first, if possible.